hy/completer.hy

```diff
@@ -0,0 +1,111 @@
+(import contextlib)
+(import os)
+(import re)
+(import sys)
+
+(import hy.macros)
+(import hy.compiler)
+(import [hy.lex.parser [mangle unmangle]])
+(import [hy.-compat [builtins string-types]])
+
+(if-python2
+  (import [--builtin-- [eval]])
+  (import [builtins [eval]]))
+
+(defmacro viewkeys [obj]
+  (if-python2
+    `(.viewkeys ~obj)
+    `(.keys ~obj)))
+
+(setv docomplete True)
+(try
+  (import readline)
+  (except [ImportError]
+    (try
+      (import pyreadline.rlmain)
+      (import pyreadline.unicode-helper)
+      (import readline)
+      (except [ImportError]
+        (setv docomplete False)))))
+
+(defclass Completer []
+  (defn --init-- [self &optional [namespace {}]]
+    (when (not (instance? dict namespace))
+      (raise (TypeError "namespace must be a dictionary")))
+
+    (setv self.namespace namespace
+          self.path [(set-comp key [key hy.compiler.-compile-table]
+                               (instance? string-types key))
+                     (viewkeys builtins.--dict--)
+                     (viewkeys (get hy.macros.-hy-macros None))
+                     (viewkeys namespace)]
+          self.tag-path [(viewkeys (get hy.macros.-hy-tag None))])
+
+    (setv module-name (.get namespace "__name__"))
+    (when module-name
+      (->> (get hy.macros.-hy-macros module-name) (viewkeys) (.append self.path))
+      (->> (get hy.macros.-hy-tag module-name) (viewkeys) (.append self.tag-path))))
```

Comment:
Why use `->>` and write these lines backwards when you could just write them forwards?